File: ruby/imperative.rb
# Imperative programming in Ruby.

# can I do lambdas, assign functions to variables in ruby?
# I think so, maybe?

puts 'simple imperative map-reduce examples'

arr = [5, 10, 15, 20, 25]
puts 'input array     : ' + arr.to_s

# anonymous methods, just call proc on object
puts "\nuse procs, anonymous methods"

# map multiply by 2
a2 = []
arr.each{ |a| a2 << a*2 }
puts 'map: mult by 2  : ' + a2.to_s

# filter if > 12
a2 = []
arr.each{ |a|
  if (a > 12) then a2 << a end
}
puts 'filter: if > 12 : ' + a2.to_s

# reduce sum
a2 = 0
arr.each{ |a| a2 += a }
puts 'reduce: sum     : ' + a2.to_s

# wrap code in named methods, keep code DRY, reusable

# map returns another array
def multBy2(a1)
  a2 = []
  a1.each{ |a| a2 << a*2 }
  return a2
end

# big = 12

# filter uses boolean
def filterBig(a1)
  big = 12
  a2 = []
  a1.each{ |a|
    if (a > big) then a2 << a end
  }
  return a2
end

# reduce returns single value
def sumArray(a1)
  a2 = 0
  a1.each{ |a| a2 += a }
  return a2
end

puts "\nuse named methods for code reuse"
puts 'map: mult by 2  : ' +  multBy2(arr).to_s
puts 'filter: if > 12 : ' + filterBig(arr).to_s
puts 'reduce: sum     : ' + sumArray(arr).to_s

File: python/imperative.py
# imperative programming in Python

arr = [5, 10, 15, 20, 25]

print 'input array:', arr

# anonymous methods
# map
a2 = []
for a in arr:
    a2.append(a * 2)

print 'map: mult by 2:', a2

big = 12

# filter
a2 = []
for a in arr:
    if a > big:
        a2.append(a)

print 'filter: bigger than', big, ':', a2

# reduce
a2 = 0
for a in arr:
    a2 += a

print 'reduce: sum:', a2

print 'named methods for code reuse'
def multBy2(ar):
    a2 = []
    for a in ar:
        a2.append(a * 2)
    return a2

def filterBig(ar):
    a2 = []
    for a in ar:
        if a > big:
            a2.append(a)
    return a2

def sumArray(ar):
    a2 = 0
    for a in arr:
        a2 += a
    return a2

print 'map: mult by 2:', multBy2(arr)
print 'filter: if >', big, ':', filterBig(arr)
print 'reduce: sum:', sumArray(arr)

# define my own functions
def mymap(func, ar):
    rr = []
    for a in ar:
        rr.append( func(a) )
    return rr

def myfilter(func, ar):
    rr = []
    for a in ar:
        if func(a):
            rr.append( a )
    return rr

def myreduce(func, ar, initval=0):
    sum = initval
    if type(ar[0]) == str:
        if type(initval) != str:
            sum = ''
    for a in ar:
        sum = func(sum, a)
    return sum

print 'mymap: mult by 2:', mymap( (lambda x: x * 2), arr)
print 'myfilter: filter >', big, ':', myfilter( (lambda x: x > big), arr)
# cases work the same as builtin map, filter
print 'myfilter: mult by 2:', myfilter( (lambda x: x * 2), arr)
print 'mymap: filter >', big, ':', mymap( (lambda x: x > big), arr)
print 'myreduce: sum:', myreduce( (lambda x,y: x + y), arr, 5)
print 'myreduce: sum str:', myreduce( (lambda x,y: x + y), ['a', 'b', 'c'])
print 'myreduce: sum str:', myreduce( (lambda x,y: x + y), ['a', 'b', 'c'], 'd')


